Multi-part form example (#39)
skuid/SamplePages: Skuid_Techniques/Multi_Part_Form/README.md

# Multi-part Form Navigation and Real-time Completion Status

> Skuid can be used to create very effective complex data-entry forms.  There are two principles that can help users work through an extensive data-entry process:  

Reviewer: ```suggestion
Skuid can be used to create very effective and complex data entry forms.  There are two principles that can help users work through an extensive data-entry process:  
```

A suggestion I'm making throughout: the more common form of this type of flow is called `data entry form` as opposed to the hyphenated `data-entry form`. My brain had to do a double take, but [the Ngram does not lie](https://books.google.com/ngrams/graph?content=data-entry+form%2C+data+entry+form&year_start=1800&year_end=2019&corpus=26&smoothing=3&direct_url=t1%3B%2Cdata%20-%20entry%20form%3B%2Cc0%3B.t1%3B%2Cdata%20entry%20form%3B%2Cc0#t1%3B%2Cdata%20-%20entry%20form%3B%2Cc0%3B.t1%3B%2Cdata%20entry%20form%3B%2Cc0). For the sake of discoverability, let's go with the more common form.

- Break up the form into chunks.
- Provide wayfinding and progress status information. 

> The example below uses the Skuid wizard component to break up a form into several different sections, but also uses the vertical navigation component - with selected-state conditions - to provide more effective wayfinding and status information for the whole form.  The example shows how to integrate the vertical navigation items and the wizard buttons so that both components stay in synch with user action. 

Reviewer: ```suggestion
The example below uses the Wizard component to break up a form into several different sections, but also uses the Vertical Navigation component - with selected-state conditions - to provide more effective wayfinding and status information for the whole form.  The example shows how to integrate the Vertical Navigation items and the Wizard buttons so that both components stay in sync with user action. 
```

The example also shows how to include dynamic information within a navigation item.  This could be a counter of completed questions, or a completion icon, etc.    

<img src="MultiPartForm.png" width="300"></img>

## Instructions
- Page API:  V2
- Data source: Only UI-only models are used. 
- Design system: [Download this Design System file](SamplePages.designsystem).  Use the Import function on the Design System page to add this system to your org.  
- Page XML:  [Copy the XML from this page](Multi_Part_Form.xml), or save it as an XML file, and upload it as a new page in your Salesforce Org.  

## Notes

> Synching the wizard and the vertical navigation selected state:

Reviewer: ```suggestion
Synching the Wizard and the Vertical Navigation selected state:
```

> - A UI Only Field is used to manage selected state in the vertical navigation component.   (See the `NavPrimary` field in the `UI Control` Model).   All navigation items and wizard buttons have action steps that update this field to an appropriate value. 

Reviewer: ```suggestion
- A UI-only feld is used to manage selected state in the Vertical Navigation component.   (See the `NavPrimary` field in the `UI Control` Model).   All navigation items and Wizard buttons have action steps that update this field to an appropriate value. 
```

Capitalization

> - The Navigation Items and the Wizard buttons each have an wizard component action that moves the user to the correct step. 

Reviewer: ```suggestion
- The navigation items and the Wizard buttons each have a Wizard component action that moves the user to the correct step. 
```

Including dynamic information in navigation step: 

> - The vertical navigation component will allow html in item labels.  This allows for formatting flexibility,  as well as the inclusion of merge syntax - to push model data into the label. 

Reviewer: ```suggestion
- The Vertical Navigation component will allow HTML in item labels.  This allows for formatting flexibility, as well as the inclusion of merge syntax to push model data into the label. 
```

> - Because the component is not connected to a particular model - this has to be global merge syntax  - like `{{$Model.FormDetail.data.0.status1}}`.

Reviewer: ```suggestion
- Because the component is not connected to a particular model this has to be global merge syntax, like `{{$Model.FormDetail.data.0.status1}}`.
```

The global merge note is less of an aside than the dashes would suggest. Let's use commas instead.

> - In order for the display within the navigation item to be updated when the data is updated - you have to call a Javscript snippet when the data is updated.   Look at the model action on the UI Control model (where the status fields are located),  as well as the Javascript found on the page. 

Reviewer: ```suggestion
- In order for the display within the navigation item to be updated when the data is updated you have to call a JavaScript snippet when the data is updated.  Look at the model action on the UI Control model (where the status fields are located),  as well as the JavaScript found on the page. 
```
